xclim/testing/tests/test_sdba/test_adjustment.py

```diff
@@ -515,6 +519,100 @@ def dist(ref, sim):
         assert (ref - scen).mean() < 5e-3
 
 
+class TestExtremeValues:
+    @pytest.mark.parametrize(
+        "c_thresh,q_thresh,frac,power",
+        [
+            ["1 mm/d", 0.95, 0.25, 1],
+            ["1 mm/d", 0.90, 1e-6, 1],
+            ["0.007 m/week", 0.95, 0.25, 2],
+        ],
+    )
+    def test_simple(self, c_thresh, q_thresh, frac, power):
+        n = 45 * 365
+
+        def gen_testdata(c, s):
+            base = np.clip(norm.rvs(loc=0, scale=s, size=(n,)), 0, None)
+            qv = np.quantile(base[base > 1], q_thresh)
+            base[base > qv] = genpareto.rvs(
+                c, loc=qv, scale=s, size=base[base > qv].shape
+            )
+            return xr.DataArray(
+                base,
+                dims=("time",),
+                coords={
+                    "time": xr.cftime_range("1990-01-01", periods=n, calendar="noleap")
+                },
+                attrs={"units": "mm/day", "thresh": qv},
+            )
+
+        ref = jitter_under_thresh(gen_testdata(-0.1, 2), 1e-3)
+        hist = jitter_under_thresh(gen_testdata(-0.1, 2), 1e-3)
+        sim = gen_testdata(-0.15, 2.5)
+
+        EQM = EmpiricalQuantileMapping(group="time.dayofyear", nquantiles=15, kind="*")
+        EQM.train(ref, hist)
+        scen = EQM.adjust(sim)
+
+        EX = ExtremeValues(c_thresh, q_thresh=q_thresh, dist="genpareto")
+        EX.train(ref, hist)
+
+        qv = (ref.thresh + hist.thresh) / 2
+        np.testing.assert_allclose(EX.ds.fit_params, [-0.1, qv, 2], atol=0.5, rtol=0.1)
+        np.testing.assert_allclose(EX.ds.thresh, qv, atol=0.15, rtol=0.01)
+
+        scen2 = EX.adjust(scen, sim, frac=frac, power=power)
+
+        # What to test???
+        # Test if extreme values of sim are still extreme
```

Comment:
My question as well. @Balinus Are you confident we're testing the right thing, here?

Reply:
Well, they way I tested it was to create a synthetic array where I manually tested what the new extreme value should be based on the cdf quantile estimations. The result of this test was then transfered .

I admit this is quite *ad hoc* and not the best way to do it. Other tests should be developed in the coming week and months as I finalize the coding (the algo will remain stable).

Reply:
Perhaps one test that could be designed is to test the bias-corrected factor for the tail parameters? I'm thinking out loud here. Perhaps it does not work.

Something like (pseudo-code)
```python
scen2 = EX.adjust(scen, sim, frac=0.01, power=power, gpparam) # 0.01 so that the transition function application domain is small and we provide gpparams here.
# sim is over the same historical period 

# Find q97 (so that we are pretty sure that the transition function of scen2 is not applied
q97 = quantile(scen2, 0.97)
exceedence = scen2[scen2>q97] - q97

gpparam_scen = gpfit( exceedence )
gpparam_scen === gpparam # triple "=", assuming it exist in Python. Or test the values of the parameters with a tolerance.
```

Does that make sense? Intuitively, the tails should be similar over the historical period.

===
Another point: does the quantile/GP parameters from the simulation are estimated over a qqmap corrected vector or from the raw/original simulation? Because the original algo is over the non-corrected values. Reading the code quickly seems like it is applied to the corrected values.

Reply:
Cool thanks! I'll try that test.

We call ""ref", "hist" and sim" the uncorrected values, "scen" is the adjusted version of "sim". So yes both distributions are fitted over uncorrected values ("ref" in the `train` and "sim` in  `adjust`). Well, if the user decides so, we're not testing anything and the first order correction happens outside `ExtremeValues`'s knowledge.

Reply:
May be something I didn't understand, but the test doesn't work... Tried it with the julia code as well to be sure, same output : very different fit parameters. But isn't that normal? Now thinking about it : the gp params are from ref, fitted on the clusters maximums with an "extreme value" threshold at quantile 0.95. Why would they be the same as gp params fitted on values of `scen` above quantile 0.97? It's not the same number of data points, no the same minimum and there is no clustering.

Reply:
(using your nomenclature)
Yes indeed, good point!

So another way would be to see if, **without** providing gp params, if you have the same gpfit over `scen` and `ref` (both with quantile 0.97). By not providing gp param, the method falls back on estimated the gp param from `ref` and thus, the tails should be similar?

But this is a long shot, since the data points are not the same and there is potential compensation between distribution's parameters.

Reply:
Well. Am I doing something wrong?

```julia
using ClimateTools

Dsim = load("xclim-testdata/sdba/fermont_CanESM2_1950-2100.nc", "pr", data_units="mm");
Dobs = load("xclim-testdata/sdba/fermont_nrcan_1950-2013.nc", "pr", data_units="mm");

ref = temporalsubset(Dobs, (1980, 1, 1), (2009, 12, 31));
hist = temporalsubset(Dsim, (1980, 1, 1), (2009, 12, 31));
sim = temporalsubset(Dsim, (2040, 1, 1), (2069, 12, 31));

scen = biascorrect_extremes(ref, hist, sim, detrend=false, window=15, rankn=50, thresnan=0.1, keep_original=false, frac=0.000000001);

refvec = ref[1].data[1,1,:];
histvec = hist[1].data[1,1,:];
scenvec = scen[1].data[1,1,:];

P = 0.95;
thres = ClimateTools.get_threshold(refvec, histvec, thres=P);

refclusters = getcluster(refvec, thres, 1.0);
GPD_ref = gpdfit(refclusters[!,:Max], threshold = thres)
# >>> Distributions.GeneralizedPareto{Float64}(μ=14.445050263404848, σ=5.912177063426245, ξ=-0.032137958612208126)

scenclusters = getcluster(scenvec, thres, 1.0)
GPD_scen = gpdfit(scenclusters[!,:Max], threshold = thres)
# >>>Distributions.GeneralizedPareto{Float64}(μ=14.445050263404848, σ=15.727722763690819, ξ=0.0)
```

Reply:
I wrote the parameters as comments after the call, here they are:
GPD_ref = `Distributions.GeneralizedPareto(μ=14.445050263404848, σ=5.912177063426245, ξ=-0.032137958612208126)`
GPD_scen = `Distributions.GeneralizedPareto(μ=14.445050263404848, σ=15.727722763690819, ξ=0.0)`.

Using 1950-2009, adjusting `hist` instead and using quantile 0.97.
GPD_ref = `Distributions.GeneralizedPareto(μ=16.861221017837533, σ=5.71010332586595, ξ=0.04983672744022191)`
GPD_scen = `Distributions.GeneralizedPareto(μ=16.861221017837533, σ=14.653519054140373, ξ=0.0)`

Reply:
Interesting. I'll take a look when I have time. Can you extract the annual maximum boxplots easily? 

I guess that since we have different data point in the end, the different parameters reflect the uncertainty of the estimation.

Reply:
Hum. So here are the box plots for `data.resample(time='YS').max()`, and all timeseries. `scen` is simple EQM (qqmap) and `scen2` is the ExtremeValues second-order adjusted data. AFAIK, the julia and python calls use the same arguments.

![image](https://user-images.githubusercontent.com/20629530/115788192-4cdbf100-a391-11eb-90eb-06d584044843.png)

Reply:
Interesting! I'll have to take a dive in the coming days and compare. The previous image showing difference around 1e-4 was based only all data over Q95? 

For the sake of simplicity, what is your code for a MWE on the Python side? I assume the Julia code is the one above. 

Thanks!

Reply:
The previous image of differences was between the results. I.e. distance between adjusted extremes in julia and python (`scen2[scen2 != scen]`). Q95 indeed.

I realize that my ref is from nrcan and you did say that extremes were badly represented there. I redid the whole thing with new data. `ref` is now AHCCD (ECCC station data, so better?). Redid the extraction for nrcan and CanESM2, for the same locations : Vancouver (BC, great rain), Kugluktuk (NU, very dry), Amos (QC, in between).

Python code:

```python
import numpy as np
import xarray as xr
import xclim as xc

from xclim import sdba
from xclim.testing import open_dataset

dsim = open_dataset('sdba/CanESM2_1950-2100.nc', branch='sdba-testfiles')
dref = open_dataset('sdba/ahccd_1950-2013.nc', branch='sdba-testfiles')

ref = xc.core.units.convert_units_to(dref.sel(time=slice('1950', '2009')).pr, 'mm/d') 
hist = xc.core.units.convert_units_to(dsim.sel(time=slice('1950', '2009')).pr, 'mm/d')

quantiles = np.linspace(0.01, 0.99, num=50)

with xr.set_options(keep_attrs=True):
    ref = ref + sdba.processing.uniform_noise_like(ref, low=1e-6, high=1e-3)
    hist = hist + sdba.processing.uniform_noise_like(hist, low=1e-6, high=1e-3)
    
EQM = sdba.EmpiricalQuantileMapping(group=sdba.Grouper('time.dayofyear', window=31), nquantiles=quantiles)
EQM.train(ref, hist)

scen = EQM.adjust(sim, interp='linear', extrapolation='constant')

EX = sdba.adjustment.ExtremeValues(cluster_thresh='1 mm/day', q_thresh=0.97)
EX.train(ref, hist)
scen2 = EX.adjust(scen, hist, frac=0.00000001)
```

Results are far more interesting!: 

![image](https://user-images.githubusercontent.com/20629530/115925992-0bf6e180-a450-11eb-9264-5415bb991854.png)

Data is here https://github.com/Ouranosinc/xclim-testdata/tree/sdba-testfiles/sdba

But I prepared a ClimateTools.jl-ready version because it is a bit more strict on the layout : 
[sdba-testing.zip](https://github.com/Ouranosinc/xclim/files/6367881/sdba-testing.zip)

Reply:
Great! So `scen2` and `scen2-JL` are quite similar then?

Reply:
Yes! They were already almost the same with the previous data. But the new one also gives more interesting results overall. Reading the codes, I can't see why the EQM/qqmap scens are so different, but that doesn't really matter here.

Reply:
Nice! 

for the qqmap aspect, I guess it is simply a matter of the choice made. I have not read deeply your code, hence can't comment on this aspect.

For the tests we mentioned, it does not make sense in the end. We can't expect to have the same GPD parameters since they are not the same (quantiles) samples.